### PR TITLE
fix: allow cluster urls to be matched

### PR DESCRIPTION
**Pull request by blakepettersson**
Related to #13646, and after discussion with @crenshaw-dev, it turns out that matching on cluster urls is not possible. This is due to the fact that the implementation of `LabelSelectorAsSelector` from `k8s.io/apimachinery` validates that a label value is no longer than 63 characters, and validates that it's alphanumeric. In order to work around that, we'll create our own implementation of `LabelSelectorAsSelector`.

This implementation has been copied verbatim, with the difference that in `isValidLabelValue`, we first check if the label value is a valid url. If it is not, we proceed with the label checks as with the original implementation.

Apart from that, the only other differences are making as much as possible to be package-private; the intent is to only make `Matches` and `LabelSelectorAsSelector` available from outside the package.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
What if, instead of a carve-out just for URLs, we simply expand our max "label" size to accommodate longer strings, like URLs?

**Comment by blakepettersson** [comment]
>What if, instead of a carve-out just for URLs, we simply expand our max "label" size to accommodate longer strings, like URLs?

@crenshaw-dev I was initially thinking just that. One issue is that the default label validation has a regex which only allows for alphanumeric characters. Of course, we could modify that to allow URL characters as well. My thinking was that in practice, most of the limits which is imposed by the label validation makes sense, _except_ for the server field. Are there any other cluster secret fields where the lifting of the length constraint would be useful?

**Comment by crenshaw-dev** [comment]
> most of the limits which is imposed by the label validation makes sense

iirc, `matchExpressions` is available on _all_ generators, rather than just cluster. Even on the cluster generator, we make annotations available as parameters. Those may be long or contain non-label characters. Other generator may produce arbitrary parameters.

I'd recommend removing all label-like restrictions. Even though we're using a tool that's meant for labels (`matchExpressions`), our input doesn't have the same restrictions that labels have.

**Comment by blakepettersson** [comment]
Alright, I dropped the restrictions on label values as per your recommendations. I'd add some documentation about this but it might be more prudent to have #13646 merged first. WDYT @crenshaw-dev?

File: applicationset/utils/selector.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+// If any of these rules is violated, an error is returned:
+// (1) The operator can only be In, NotIn, Equals, DoubleEquals, Gt, Lt, NotEquals, Exists, or DoesNotExist.
+// (2) If the operator is In or NotIn, the values set must be non-empty.
+// (3) If the operator is Equals, DoubleEquals, or NotEquals, the values set must contain one value.
+// (4) If the operator is Exists or DoesNotExist, the value set must be empty.
+// (5) If the operator is Gt or Lt, the values set must contain only one value, which will be interpreted as an integer.
+// (6) The key is invalid due to its length, or sequence
+//
+//	of characters. See validateLabelKey for more details.
+//
```
Comment (crenshaw-dev) [inline applicationset/utils/selector.go:154]: I actually don't think we need to validate the key, either. But this PR is a move in the right direction. Merging, we can always address key validation later.

**Review by crenshaw-dev — APPROVED** [review]

**Comment by crenshaw-dev** [comment]
@blakepettersson do you mind setting up the cherry-pick PRs?

**Pull request merged**